### ISSUE TITLE
Node with larger number of shards was not added to cluster if first node have less number of shards

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylladb#12252]
## Installation details

Cluster size: 4 nodes (i3.4xlarge)

Scylla Nodes used in this run:
- longevity-50gb-12h-master-db-node-5981775c-4 (34.241.187.47 | 10.4.0.84) (shards: -1) - Node was not initialized
- longevity-50gb-12h-master-db-node-5981775c-3 (3.253.23.51 | 10.4.0.191) (shards: -1)  - Node was not initialized
- longevity-50gb-12h-master-db-node-5981775c-2 (3.253.48.130 | 10.4.3.154) (shards: 14)
- longevity-50gb-12h-master-db-node-5981775c-1 (18.203.223.233 | 10.4.3.248) (shards: 11)

OS / Image: `ami-063cdd564cd2fbe46` (aws: eu-west-1)

Test: `longevity-50gb-asymmetric-cluster-12h-test`
Test id: `5981775c-d8d1-4725-add3-c4afe33b8471`
Test name: `scylla-master/raft/longevity-50gb-asymmetric-cluster-12h-test`
Test config file(s):

- [longevity-150GB-12h-autorization-LimitedMonkey.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/9dd49aca98a05dcf3322aa525bd15abc2307f932/test-cases/longevity/longevity-150GB-12h-autorization-LimitedMonkey.yaml)

## Issue description

Cluster initialization failed when number of shards is different on node. Node1 was started with --smp=11, Node2 was started with --smp=14. And scylla on node was stucked and not initialized.

node1:
```
2022-12-08T19:17:53+00:00 longevity-50gb-12h-master-db-node-5981775c-1     !INFO | scylla[5477]: command used: "/usr/bin/scylla --blocked-reactor-notify-ms 25 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-internal-error 1 --abort-on-ebadf 1 --enable-sstable-key-validation 1 --smp 11 --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 1-7,9-15 --lock-memory=1"


2022-12-08T19:18:22+00:00 longevity-50gb-12h-master-db-node-5981775c-1     !INFO | scylla[5477]:  [shard  0] init - Scylla version 5.2.0~dev-0.20221208.a076ceef97d5 initialization completed.
```
node2:
```
2022-12-08T19:20:46+00:00 longevity-50gb-12h-master-db-node-5981775c-2     !INFO | scylla[5524]: Scylla version 5.2.0~dev-0.20221208.a076ceef97d5 with build-id 020ec076898a692651fd48edfb1920fc190cd81e starting ...
2022-12-08T19:20:46+00:00 longevity-50gb-12h-master-db-node-5981775c-2     !INFO | scylla[5524]: command used: "/usr/bin/scylla --blocked-reactor-notify-ms 25 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-internal-error 1 --abort-on-ebadf 1 --enable-sstable-key-validation 1 --smp 14 --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 1-7,9-15 --lock-memory=1"

2022-12-08T19:20:48+00:00 longevity-50gb-12h-master-db-node-5981775c-2     !INFO | scylla[5524]:  [shard  0] init - starting Raft Group Registry service


```

And scylla was not fully started.

- Restore Monitor Stack command: `$ hydra investigate show-monitor 5981775c-d8d1-4725-add3-c4afe33b8471`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=5981775c-d8d1-4725-add3-c4afe33b8471)
- Show all stored logs command: `$ hydra investigate show-logs 5981775c-d8d1-4725-add3-c4afe33b8471`

## Logs:
- **db-cluster-5981775c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/db-cluster-5981775c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/db-cluster-5981775c.tar.gz)
- **monitor-set-5981775c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/monitor-set-5981775c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/monitor-set-5981775c.tar.gz)
- **loader-set-5981775c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/loader-set-5981775c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/loader-set-5981775c.tar.gz)
- **sct-runner-5981775c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/sct-runner-5981775c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/5981775c-d8d1-4725-add3-c4afe33b8471/20221208_202248/sct-runner-5981775c.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/raft/job/longevity-50gb-asymmetric-cluster-12h-test/2/)

**Comment by roydahan** [comment]
@aleksbykov it looks like you mixed information from both the jepsen test and the asymmetric longevity.
Please clarify it and possibly separate them into 2 different comments (with their own logs).

**Comment by DoronArazii** [comment]
Removing triage since Kostja's team took it

**Comment by avikivity** [comment]
Fix present on 5.2. Not backporting to 5.1 where raft is still experimental and the code is very different.
